### PR TITLE
Add `toBns` and allow for number arguments

### PR DESCRIPTION
**Pull request by samholmes**
### CHANGELOG

- added: New `toBns` function to convert JS number to big number strings
- changed: Implement `toBns` for all public API to expand argument type support

### Dependencies

 none

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205419647873859

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import BN from 'bn.js'
 
 // const MAX_DECIMALS = 10
+const EXPO_REGEX = /e((?:\+|-)\d+)$/
```
Comment (samholmes) [inline src/index.ts:8]: call this `SCI_NOTATION_REGEX`.

File: src/index.ts [hunk 2]
```diff
@@ -224,6 +224,14 @@ function div(
   return base === 10 ? out : out.replace(/^(-)?/, '$10x')
 }
 
+function E(x1: string, exponent: number): string {
```
Comment (samholmes) [inline src/index.ts:228]: Don't export this one. Name it `sciNotation`

File: src/index.ts [hunk 3]
```diff
@@ -409,6 +410,20 @@ const ceil = (x1: string, precision: number): string =>
 const round = (x1: string, precision: number): string =>
   precisionAdjust('round', x1, precision)
 
+function toBigNumberString(n: number | string): string {
```
Comment (samholmes) [inline src/index.ts:413]: Change `toBns(n: number | string)`.

Add a `fixSciNotation` which does the sci-notation part and use that on all string inputs for each function that  deals with a BNS

**Review by samholmes — COMMENTED** [review]
from review-bar with paul

File: src/index.ts [hunk 4]
```diff
@@ -401,6 +410,20 @@ const ceil = (x1: string, precision: number): string =>
...
   precisionAdjust('round', x1, precision)
 
+function toBns(n: number | string): string {
+  let out = typeof n === 'number' ? n.toString() : n
+  // Handle scientific notation
+  const match = out.match(SCI_NOTATION_REGEX)
+
+  if (match != null) {
+    const exponent = parseInt(match[1])
+    const base = out.replace(SCI_NOTATION_REGEX, '')
```
Comment (paullinator) [inline src/index.ts]: Replace with regex with `SCI_NOTATION_REGEX = /(.[0-9]*)e((?:\+|-)\d+)$/` Then you only have to apply it once to get both the exponent and base.

**Review by paullinator — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 5]
```diff
@@ -471,8 +471,9 @@ function precisionAdjust(
 }
 
 function sciNotation(x1: string, exponent: number): string {
+  console.log(x1, exponent)
```
Comment (paullinator) [inline src/index.ts:474]: stray log

File: test/biggystring.test.ts [hunk 6]
```diff
@@ -199,12 +199,33 @@ describe('toBns', function () {
...
+    assert.equal(toBns(bigStr), '3000000000000000000000')
+  })
+  it('scientific notation: small numbers', function () {
+    const small = 0.000000000000000000003
+    const smallStr = small.toString()
+    assert.equal(smallStr, '3e-21')
+    assert.equal(toBns(small), '0.000000000000000000003')
+    assert.equal(toBns(smallStr), '0.000000000000000000003')
+  })
+  it('scientific notation: big negative numbers', function () {
```
Comment (paullinator) [inline test/biggystring.test.ts:220]: I still feel like this needs more tests. 

Checking decimal values like `3.1234e12` `.001e12` `0.000123e12`

invalid values like `3.1.0e123` `3x0e12` `

**Review by paullinator — CHANGES_REQUESTED** [review]

**Review by paullinator — CHANGES_REQUESTED** [review]
more tests

File: src/index.ts [hunk 7]
```diff
@@ -223,6 +223,8 @@ export function toBns(n: number | string): string {
     out = sciNotation(base, exponent)
   }
 
+  validate(out)
```
Comment (paullinator) [inline src/index.ts:226]: This is redundant for most operations since they already call validate. Create a toBnsInner that gets used by other ops.

**Review by paullinator — CHANGES_REQUESTED** [review]
One more small change and we're good. new tests look great.

**Review by paullinator — APPROVED** [review]

**Pull request merged**